pointer masking support (#368)
riscv/riscv-cheri: src/ptrmask-integration.adoc

[#section_pointer_masking_integration]
== Integrating {cheri_base_ext_name} and {cheri_default_ext_name} with Pointer Masking

> The pointer masking extensions SSnpm, Sspm and Supm are compatible with {cheri_default_ext_name}.

Reviewer: According to the RISC-V Pointer Masking spec, the existing Pointer Masking extensions are:
* Smmpm
* Smnpm
* Ssnpm

Could we align this text with the contents of the RISC-V spec?

Author: I took the names from https://github.com/riscv/riscv-profiles/blob/main/src/rva23-profile.adoc

Author: ```suggestion
The pointer masking extensions Smmpm, Smnpm, SSnpm, Sspm and Supm are compatible with {cheri_default_ext_name}.
```

In {cheri_int_mode_name} all data addresses are integer pointers, which
are interpreted as being XLEN-wide, and may be subject to pointer masking.
All data accesses are checked against <<ddc>> which is unaffected by
pointer masking. Therefore no capability bounds encoding is affected.

In {cheri_cap_mode_name} all pointers are capabilities, and the bounds

> encoding relies on XLEN-wide addresses, and therefore the definitions

Author: ```suggestion
encoding relies on XLEN-wide addresses, and therefore the definitions of the bounds
```

> will need to be updated to be usable for pointer masking.

Reviewer: Which definitions are you referring to?

> Therefore supporting pointer masking in {cheri_cap_mode_name} is future work.

Author: ```suggestion
Therefore supporting pointer masking in {cheri_cap_mode_name} is future work, and so
Xenfcfg.PMM is currently ignored in {cheri_cap_mode_name}.
```

[#section_pointer_masking_integration]
== Integrating {cheri_base_ext_name} and {cheri_default_ext_name} with Pointer Masking

The pointer masking extensions Smmpm, Smnpm, SSnpm, Sspm and Supm are compatible with {cheri_default_ext_name}.

In {cheri_int_mode_name} all data addresses are integer pointers, which
are interpreted as being XLEN-wide, and may be subject to pointer masking.
All data accesses are checked against <<ddc>> which is unaffected by
pointer masking. Therefore no capability bounds encoding is affected.

> In {cheri_cap_mode_name} all pointers are capabilities. If pointer masking is enabled

Reviewer 2: This is the way that Morello did it, but they faced some unintended consequences as a result. Because of how sign extension would change bounds decode, modifying a capability address (completely in bounds) can result in dramatically different bounds. This was even a bug for a while as it beat the fast representability check. They ended up having to special case changing the 56th bit of address, if the capability had a specific range of E's, to always untag when modified even though the capability was never taken out of bounds.

I think we should probably learn from this lesson that whilst the address should eventually be sign extended for the purpose of generating a vaddr for the next stage of memory fetch, sign extension should likely not be a part of bounds calculation / bounds check so that in-bounds capabilities are always representable.

I would suggest for the purpose of bounds decode / bounds check high bits just be masked (not extended)

Reviewer 3: Agreed we need to fix this for a proper integration, but for now just doing the bare minimum makes the most sense.

encoding relies on XLEN-wide addresses, and therefore the definitions of the bounds
will need to be updated to be usable for pointer masking.

> Therefore supporting pointer masking in {cheri_cap_mode_name} is future work, and so

Reviewer 3: I'm not sure this is ideal since it means you also have to mask the PMM bit whenever you do a modesw in hybrid code? I think it would be more consistent to always mask off the bits and just feed those into the bounds check, and rephrase this as "pointer masking is not applied to capability decoding"? 

Setting/clearing the high bits could create an unrepresentable capability (which is why Morello sort of ignored them) and you can't use them for metadata unless you have an almighty capability. As long as the bounds are still decoded relative to the address stored in the capability rather than the masked address, masking can't change the meaning of capabilities. And since it just changes inaccessible virtual addresses I don't think masking prior to the bounds check is a problem? It's not particularly useful but at least it's more consistent than forcing PMM off?

Author: ok - I've changed it - please have another look

[#section_pointer_masking_integration]
== Integrating {cheri_base_ext_name} and {cheri_default_ext_name} with Pointer Masking

The pointer masking extensions Smmpm, Smnpm, SSnpm, Sspm and Supm are compatible with {cheri_default_ext_name}.

In {cheri_int_mode_name} all data addresses are integer pointers, which
are interpreted as being XLEN-wide, and may be subject to pointer masking.
All data accesses are checked against <<ddc>> which is unaffected by
pointer masking. Therefore no capability bounds encoding is affected.

> In {cheri_cap_mode_name} all pointers are capabilities. If pointer masking is enabled
> then the entire address field, including any bits representing the pointer mask, are used
> for bounds calculation. The dereferenced address has the masked bits replaced by sign

Reviewer 3: ```suggestion
For instructions using capabilities (e.g. loads/stores in {cheri_cap_mode_name}), the final access address is subject to pointer masking, but the computed bounds are not. The entire address field, including any bits representing the pointer mask, are used
for bounds calculation. When pointer masking is enabled, the dereferenced address has the masked bits replaced by sign
```

maybe something like this?

extension before the bounds check.

NOTE: This scheme doesn't seem very useful, but the problem is the dynamic configuration

>  of pointer masking which can dynamically update meaning of the address within the capability,

Reviewer: ```suggestion
 of pointer masking which can arbitrarily update the meaning of the address within the capability,
```

>  and so the full address field must be used to calculate bounds.

Reviewer: ```suggestion
 so the full address field must be used to calculate bounds.
```

>  There is future work required to determine a more useful way of applying pointer masking in
>  {cheri_cap_mode_name}.

Reviewer 3: ```suggestion
 There is future work required to determine a more useful way of applying pointer masking to capabilities.
```
